crates/flowctl/src/source/unbundle.rs

```diff
@@ -0,0 +1,338 @@
+/// as file(s) in the given `dir_path`. The input catalog is typically one that
+/// is bundled, but this is not a requirement. If the given catalog contains
+/// inlined resources, then they will be written out as separate files and the
+/// spec will have them replaced with relative paths to them. Resources that
+/// are _not_ inlined will be left unmodified. Note that `unbundle` does _not_
+/// put things into nested subdirectories. The entire catalog will be written
+/// directly within the given directory.
+pub async fn unbundle(
+    new_catalog: models::Catalog,
+    dir_path: &Path,
+    spec_filename: &str,
+    json: bool,
+    existing: Existing,
+) -> anyhow::Result<()> {
+    fs::create_dir_all(dir_path).await?;
+    let flow_yaml_path = dir_path.join(spec_filename);
+
+    let maybe_catalog = resolve_catalog_to_write(&flow_yaml_path, new_catalog, existing).await?;
+    let Some(mut new_catalog) = maybe_catalog else {
+        return Ok(());
+    };
+
+    let serializer = Serializer { json };
+    let ext = if json { "json" } else { "yaml" };
+
+    for (name, collection) in new_catalog.collections.iter_mut() {
+        let base = base_name(name);
+
+        // Potentially write out separate schema files for the collection
+        // schema / write_schema / read_schema.
+        for (schema, filename) in [
+            (collection.schema.as_mut(), format!("{base}.schema.{ext}")),
+            (
+                collection.write_schema.as_mut(),
+                format!("{base}.write.schema.{ext}"),
+            ),
+            (
+                collection.read_schema.as_mut(),
+                format!("{base}.read.schema.{ext}"),
+            ),
+        ] {
+            let Some(schema) = schema else { continue };
+
+            maybe_indirect_schema(dir_path, &filename, &serializer, schema, existing).await?;
+        }
+
+        if let Some(derivation) = &mut collection.derivation {
+            maybe_indirect_schema(
+                dir_path,
+                &format!("{base}.register.schema.{ext}"),
+                &serializer,
+                &mut derivation.register.schema,
+                existing,
+            )
+            .await?;
+
+            if let Some(typescript) = &mut derivation.typescript {
+                // Indirect the TypeScript module to a file, and track its reference.
+                let ts_filename = format!("{base}.ts");
+                let file_path = dir_path.join(&ts_filename);
+
+                write_file(&file_path, typescript.module.as_bytes(), existing).await?;
+                typescript.module = ts_filename;
+            }
+        }
+    }
+
+    for (name, capture) in new_catalog.captures.iter_mut() {
+        let base = base_name(name);
+
+        if let models::CaptureEndpoint::Connector(connector) = &mut capture.endpoint {
+            // Indirect the connector config to a file, and track its reference.
+            let filename = format!("{base}.config.{ext}");
```

Comment:
Worth calling these `{base}.capture.{ext}` and `{base}.materialization.{ext}`?

Reply:
I opted to keep these exactly the same as they are in `draft develop` so that existing repos don't have to deal with files being re-named. On a related note, there's no deletion of existing "unbundled" files like schemas and endpoint configs, so changing the names here would potentially result in both `{base}.capture.{ext}` and `{base}.config.{ext}` existing side by side.

Reply:
Mhmmmmm, I didn't consider the backwards compatibility aspect. Makes sense!